File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/articles/ArticlesViewModel.java
package com.andruid.magic.newsdaily.articles;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.andruid.magic.newsloader.model.News;

import static com.andruid.magic.newsloader.data.Constants.PAGE_SIZE;

public class ArticlesViewModel extends ViewModel {
    private MutableLiveData<String> queryLiveData;
    private LiveData<PagedList<News>> pagedListLiveData;

    ArticlesViewModel(String language) {
        queryLiveData = new MutableLiveData<>();
        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(PAGE_SIZE)
                .build();
        pagedListLiveData = Transformations.switchMap(queryLiveData, query ->
                new LivePagedListBuilder<>(new ArticlesDataSourceFactory(language,
                        queryLiveData.getValue()), pagedListConfig).build());
    }


    public LiveData<PagedList<News>> loadArticles(String query) {
        queryLiveData.postValue(query);
        return pagedListLiveData;
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/articles/ArticlesViewModel.kt
package com.andruid.magic.newsdaily.articles

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import androidx.lifecycle.ViewModel
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.andruid.magic.newsloader.data.Constants
import com.andruid.magic.newsloader.model.News

class ArticlesViewModel(private val language : String) : ViewModel() {
    private val queryLiveData: MutableLiveData<String> = MutableLiveData()
    private var pagedListLiveData : LiveData<PagedList<News>>

    init {
        val pagedListConfig = PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .build()
        pagedListLiveData = Transformations.switchMap(queryLiveData) { query: String? ->
            LivePagedListBuilder(ArticlesDataSourceFactory(language,
                    query ?: ""), pagedListConfig).build()
        }
    }

    fun setQuery(query : String) { queryLiveData.postValue(query) }
}